Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoEntityOverrides.cpp

```diff
@@ -0,0 +1,145 @@
+                PrefabDom entityDomAfterUpdate;
+                m_instanceToTemplateInterface->GenerateDomForEntity(entityDomAfterUpdate, *entity);
+
+                // Get the before-state entity DOM inside the source template DOM of the top instance.
+                const TemplateId topTemplateId = link->get().GetSourceTemplateId();
+                const PrefabDom& topTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom(topTemplateId);
+
+                PrefabDom overridePatches;
+                {
+                    // This scope is added to limit their usage and ensure DOM is not modified when it is being used.
+                    // DOM value pointers can't be relied upon if the original DOM gets modified after pointer creation.
+                    const PrefabDomValue* entityDomInTopTemplate = entityDomPathFromTopInstance.Get(topTemplateDom);
+
+                    if (entityDomInTopTemplate)
+                    {
+                        m_instanceToTemplateInterface->GeneratePatch(overridePatches, *entityDomInTopTemplate, entityDomAfterUpdate);
+                    }
+                    else
+                    {
+                        // Merge patches to an entity DOM as add-entity override.
+                        // Path is empty because there is no subpath for add-entity override patch like "/Components/Component_[123]".
+                        overridePatches.SetArray();
+                        PrefabUndoUtils::AppendAddEntityPatch(overridePatches, entityDomAfterUpdate, "");
+                    }
+                }
+
+                PrefabOverridePrefixTree overrideSubTree = link->get().GenerateOverrideSubTreeForEntity(
+                    overridePatches, entityPathFromTopInstance);
+                m_overrideSubTrees[AZ::Dom::Path(entityPathFromTopInstance)] = AZStd::move(overrideSubTree);
+
+                // Preemptively updates the cached DOM to prevent reloading instance DOM.
+                if (cachedOwningInstanceDom.has_value())
+                {
+                    PrefabUndoUtils::UpdateEntityInInstanceDom(cachedOwningInstanceDom, entityDomAfterUpdate, entityAliasPath);
+                }
+            }
+        }
+
+        void PrefabUndoEntityOverrides::Undo()
+        {
+            UpdateLink();
+        }
+
+        void PrefabUndoEntityOverrides::Redo()
+        {
+            UpdateLink();
+        }
+
+        void PrefabUndoEntityOverrides::UpdateLink()
+        {
+            LinkReference link = m_prefabSystemComponentInterface->FindLink(m_linkId);
+            if (link.has_value())
+            {
+                for (auto& [pathToSubTree, overrideSubTree] : m_overrideSubTrees)
+                {
+                    PrefabOverridePrefixTree prevSubTree = AZStd::move(link->get().RemoveOverrides(pathToSubTree));
+                    link->get().AddOverrides(pathToSubTree, AZStd::move(overrideSubTree));
+                    overrideSubTree = AZStd::move(prevSubTree);
```

Comment:
It took me a while to understand that you are using the same map to store 'before' trees when redoing and 'after' states when undoing. Maybe add a comment in the header file to explain the usage of the map

Reply:
Yeah. `m_overrideSubTrees` would be confusing. Better to rename it to `m_prevSubTrees` and add comments.